src/hooks/useDataFilter.tsx

```diff
@@ -0,0 +1,65 @@
+import { useMemo, useState } from 'react'
+import { useDebounce } from './useDebounce'
+
+interface HookResults<T> {
+  filteredData: T[]
+  search: string
+  showFilter: boolean
+  handleSearch: (e: React.ChangeEvent<HTMLInputElement>) => void
+  handleToggleFilter: (e: React.ChangeEvent<HTMLInputElement>) => void
+  clearFilters: () => void
+}
+
+export interface HookParams<T> {
+  data: T[]
+  searchDebounceTime?: number
+  isSearchFilter?: boolean
+  customStopCheck?: (...any: unknown[]) => boolean
+  filterFnFactory: (searchTxt: string) => (params: T, index: number, array: T[]) => boolean
+}
+
+function useDataFilter<T>({
+  data,
+  searchDebounceTime = 500,
+  isSearchFilter = true,
+  customStopCheck,
+  filterFnFactory,
+}: HookParams<T>): HookResults<T> {
+  const [search, setSearch] = useState('')
+  // searchFilter prop turns on/off by default. If using search filter, we dont want to use toggle
+  const [showFilter, setShowFilter] = useState(isSearchFilter)
+  const { value: debouncedSearch, setImmediate: setDebouncedSearch } = useDebounce(search, searchDebounceTime)
+
+  const handleSearch = (e: React.ChangeEvent<HTMLInputElement>): void => setSearch(e.target.value)
+  const handleToggleFilter = (): void => setShowFilter(state => !state)
+  const clearFilters = (): void => {
+    setSearch('')
+    setDebouncedSearch('')
+    setShowFilter(false)
+  }
```

Comment:
Memoize together, as setState are stable refs.
It's nicer to have these functions as const when passing to components later

Reply:
you mean like:
```ts
const handlers = useMemo(() => {
    const handleSearch = (e: React.ChangeEvent<HTMLInputElement>): void => setSearch(e.target.value)
    const handleToggleFilter = (): void => setShowFilter(state => !state)
    const clearFilters = (): void => {
      setSearch('')
      setDebouncedSearch('')
      setShowFilter(false)
    }

    return {
      handleSearch,
      handleToggleFilter,
      clearFilters,
    }
  }, [setDebouncedSearch])
```
? or  just the clear filters

Reply:
what do you mean by easier to pass as `const`s?

Reply:
exactly like you described, `handlers` 👍 
`const` worded awkwardly, I meant stable references from `handlers`

Reply:
thanks man